### DISCUSSION TITLE
`aws-iam-authenticator` for baremetal variant

### DISCUSSION THREAD

**Discussion by StefanTheWiz** [bottlerocket-os/bottlerocket#2809] [Q&A] [answered]
Hi!

I'm trying to use AWS EKS as the control plane with bottlerocket baremetal nodes.
I was wondering if there was a particular reason why `aws-iam-authenticator` is not included in baremetal variant?

Based on the [README](https://github.com/bottlerocket-os/bottlerocket/blob/develop/README.md), I was kinda expecting that it would be included (for instance for using IAM Roles Anywhere).
> To start, we're focusing on the use of Bottlerocket as a host OS in AWS EKS Kubernetes clusters and Amazon ECS clusters.

As an experiment, I've built a version of baremetal that has the `aws-iam-authenticator` (diff [here](https://github.com/StefanTheWiz/bottlerocket/commit/65f1737fd37cbe0d12cb33a59de65a021cc37cc1)).
However, the update process might remove the authenticator, which might mean maintaining our own TUF repo and that would be a tough sell.

**Comment by rpkelly** [comment]
Hi Stefan!

I believe we chose not to include `aws-iam-authenticator` on the baremetal variants for two main reasons. 1) We didn't want to assume that they would be used with EKS, and 2) since we weren't assuming EKS, we decided to save some space on the image size.

It is possible to use EKS without `aws-iam-authenticator` via the `get-token` [api](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/eks/get-token.html)

**Reply by StefanTheWiz** [reply]
> I believe we chose not to include aws-iam-authenticator on the baremetal variants for two main reasons. 1) We didn't want to assume that they would be used with EKS

Yes, not everyone might use BR with EKS. However, I couldn't find anything in the repo stating that bare metal doesn't support `authentication-mode: "aws"`?
If the intention is that the AWS authentication mode is supported across all variants, then the `aws-iam-authenticator` package should be included.

This is because the kubernetes package has this [kubelet-kubeconfig](https://github.com/bottlerocket-os/bottlerocket/blob/113b49ba27de11dbe2cc56130fe52efeeb8beaf3/packages/kubernetes-1.24/kubelet-kubeconfig#L19-L34) for bootstrap, which doesn't safeguard from the `aws-iam-authenticator` being used on non-AWS variants (where the executable is missing).

So one option could be to class this as a bug, with the resolution of either:
* checking that `variant == "aws"` in the kubeconfig template; or 
* including the `aws-iam-authenticator` package (I'm clearly favouring this option 😇)

---
> 2 ) since we weren't assuming EKS, we decided to save some space on the image size.

Looking at the [`aws-iam-authenticator` releases page](https://github.com/kubernetes-sigs/aws-iam-authenticator/releases), the binaries are around 50MB in size, which seems pretty small given the whole OS takes about 5GB.

Furthermore, bare metal storage is generally cheap. For instance, [Hetzner](https://www.hetzner.com/additional-hardware/) charges between €0.012-0.019 / GB / month. A 32GB SD card for Raspberry Pi is just £6, which works out at about €0.21/GB.

I can't speak for other BR users but, personally, I'm not very concerned about saving 50 MB storage.


---
> It is possible to use EKS without aws-iam-authenticator via the get-token [api](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/eks/get-token.html)

As far as I can tell, none of the variants come with AWS CLI installed? 🤔
I might be able to use a bootstrap container to get the EKS token and inject it in `/etc/kubernetes/kubelet/kubeconfig` but I'll wait, pending on the comments from (1) above.

**Reply by rpkelly** [reply, marked as answer]
I've created issue #2823 regarding the `aws-iam-authenticator` as well as opened an issue [here](https://github.com/bottlerocket-os/project-website/issues/36) regarding documentation.

I think I understand your use case better now. Adding the package may still require some more discussion though, and seems like a great topic for an upcoming [community meeting](https://www.meetup.com/bottlerocket-community/).

It would also be great to get feedback from anyone else looking to use `aws-iam-authenticator` this way. Please add any comments to help understand the need better.

**Reply by StefanTheWiz** [reply]
Thanks @rpkelly. I'll join the community meeting tomorrow.
